### ISSUE TITLE
Off main thread symbolic evaluation

### ISSUE THREAD

**Issue by ezyang** [facebookresearch/torchdynamo#138]
Standard JIT design has compilation run off-thread, so that normal user code can continue running while compilation is still occurring. Unfortunately, torchdynamo's implementation in Python makes this difficult to do, as in standard Python there is a GIL which prevents useful parallelization of Python work. This means that we need to think about our off main thread strategy earlier rather than later, as it may imply architectural changes that are easier to do earlier in this project's lifetime.

I can imagine a number of possible ways to skin this cat:

* Put torchdynamo evaluation in a separate process
* Put torchdynamo in a subinterpreter in the same process
  * All strategies involving a separate Python interpreter will have to somehow transmit all of the frame information as well as locals for evaluation
* Rewrite torchdynamo in C++ (ugh)
* Rewrite torchdynamo in RPython
* Do nothing and pray the host program spends enough time in PyTorch operations that we can make useful progress with standard threading
* Do nothing for symbolic evaluation, but offload FX graph compilation to another thread

It would be good to know the PoR here.

**Comment by ezyang** [comment]
One argument for NOT running symbolic evaluation off thread is that most of our optimization backends introduce numerical differences (they are not bit-for-bit compatible); thus, to get predictable behavior, we must either ALWAYS apply the optimized version or NEVER apply it (and not some nondeterministic mix between the two.)

**Comment by anijain2305** [comment]
@jansel Adding a ping in case this is relevant now, given TorchInductor (and sometimes even nvfuser) can take a large time to compile.

**Comment by jansel** [comment]
Yes, this is still relevant.  Compile times are getting up there...
